File: usvm-ts/src/test/kotlin/org/usvm/samples/StaticFields.kt
package org.usvm.samples

import org.jacodb.ets.model.EtsScene
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import org.usvm.api.TsTestValue
import org.usvm.util.TsMethodTestRunner

@Disabled("Statics are not fully supported, yet")
class StaticFields : TsMethodTestRunner() {

    private val className = this::class.simpleName!!

    override val scene: EtsScene = loadSampleScene(className)

    @Test
    fun `test static access get`() {
        val method = getMethod("StaticNumber", "getValue")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r -> r.number == 10.0 },
        )
    }

    @Test
    fun `test static default value`() {
        val method = getMethod("StaticDefault", "getValue")
        discoverProperties<TsTestValue>(
            method = method,
            { r -> r == TsTestValue.TsUndefined },
        )
    }

    @Test
    fun `test static modification`() {
        val method = getMethod("StaticModification", "incrementTwice")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r -> r.number == 2.0 },
        )
    }

    @Test
    fun `test static inheritance`() {
        val method = getMethod("StaticDerived", "getId")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r -> r.number == 142.0 },
        )
    }

    @Test
    fun `test static inheritance shadowing parent`() {
        val method = getMethod("StaticChild", "getParentId")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r -> r.number == 100.0 },
        )
    }

    @Test
    fun `test static inheritance shadowing child`() {
        val method = getMethod("StaticChild", "getChildId")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r -> r.number == 200.0 },
        )
    }

    @Test
    fun `test static inheritance shadowing`() {
        val method = getMethod("StaticChild", "getId")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r -> r.number == 200.0 },
        )
    }

    @Test
    fun `test static boolean toggle`() {
        val method = getMethod("StaticBoolean", "toggleAndGet")
        discoverProperties<TsTestValue.TsBoolean>(
            method = method,
            { r -> r.value == true },
        )
    }

    @Disabled("Array::push() is not supported")
    @Test
    fun `test static array modification`() {
        val method = getMethod("StaticArray", "pushTwice")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r -> r.number == 5.0 },
        )
    }

    @Disabled("Sort mismatch on union type")
    @Test
    fun `test static null initialization`() {
        val method = getMethod("StaticNull", "initialize")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r -> r.number == 5.0 },
        )
    }

    @Disabled("Statics are hard... See issue 607 in AA")
    @Test
    fun `test static object manipulation`() {
        val method = getMethod("StaticObject", "modifyAndGet")
        discoverProperties<TsTestValue.TsClass>(
            method = method,
            { r ->
                (r.properties["enabled"] as TsTestValue.TsBoolean).value == true &&
                    (r.properties["count"] as TsTestValue.TsNumber).number == 15.0
            },
        )
    }

    @Test
    fun `test static access sum`() {
        val method = getMethod("StaticAccess", "calculateSum")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r -> r.number == 15.0 },
        )
    }

    @Disabled("Array length cannot be properly read from memory due to array type mismatch")
    @Test
    fun `test static access swap`() {
        val method = getMethod("StaticAccess", "swapAndGetValues")
        discoverProperties<TsTestValue.TsArray<TsTestValue.TsNumber>>(
            method = method,
            { r -> r.values.map { it.number } == listOf(2.0, 1.0) },
        )
    }

    @Test
    fun `test static any type`() {
        val method = getMethod("StaticAny", "getNumber")
        discoverProperties<TsTestValue.TsNumber>(
            method = method,
            { r -> r.number == 10.0 },
        )
    }
}

File: usvm-ts/src/test/resources/samples/StaticFields.ts
// @ts-nocheck
// noinspection JSUnusedGlobalSymbols

// Test: Basic number static
class StaticNumber {
    static value = 10;

    static getValue(): number {
        return this.value;
    }
}

// Test: Default value initialization
class StaticDefault {
    static value: number;

    static getValue(): number {
        return this.value;
    }
}

// Test: Sequential static modifications with value persistence
class StaticModification {
    static count = 0;

    static incrementTwice(): number {
        this.count += 1;
        this.count += 1;
        return this.count;
    }
}

// Test: Inheritance
class StaticBase {
    static id = 100;
}

class StaticDerived extends StaticBase {
    static f = 42;

    static getId(): number {
        return this.id + this.f;
    }
}

// Test: Inheritance shadowing
class StaticParent {
    static id = 100;
}

class StaticChild extends StaticParent {
    override static id = 200;

    getParentId(): number {
        return StaticParent.id;
    }

    getChildId(): number {
        return StaticChild.id;
    }

    static getId(): number {
        return this.id;
    }
}

// Test: Boolean static toggle
class StaticBoolean {
    static flag: boolean = false;

    static toggleAndGet(): boolean {
        this.flag = !this.flag;
        return this.flag;
    }
}

// Test: Array static manipulation
class StaticArray {
    static numbers = [1, 2, 3];

    static pushTwice(): number {
        this.numbers.push(4);
        this.numbers.push(5);
        return this.numbers.length;
    }
}

// Test: Null initialization and update
class StaticNull {
    static value: number | null = null;

    static initialize(): number {
        this.value = 5;
        return this.value;
    }
}

// Test: Object static operations
class StaticObject {
    static config: Config = {enabled: true, count: 10};

    static modifyAndGet(): Config {
        this.config.increment();
        this.config.increment();
        this.config.increment();
        this.config.flip();
        this.config.increment();
        this.config.increment();
        this.config.flip();
        return this.config;
    }
}

class Config {
    enabled: boolean;
    count: number;

    flip(): void {
        this.enabled = !this.enabled;
    }

    increment(): void {
        this.count += 1;
    }
}

// Test: Field swapping
class StaticAccess {
    static a = 5;
    static b = 10;

    static calculateSum(): number {
        return this.a + this.b;
    }

    static swapAndGetValues(): number[] {
        [this.a, this.b] = [this.b, this.a];
        return [this.a, this.b];
    }
}

// Test: Any-type static
class StaticAny {
    static value: any = 10;

    static getNumber(): number {
        return this.value;
    }
}
